### PR TITLE
Use Interchange 0.4

### PR DESCRIPTION
**Pull request by mattwthompson**

### CODE REVIEW

File: devtools/conda-envs/test_env.yaml [hunk 1]
```diff
@@ -10,9 +10,9 @@ dependencies:
   - numpy
   - openmm
   - openff-toolkit =0.16
-  - openff-interchange ~=0.3.25
+  - openff-interchange ~=0.4.0rc1
   - openff-utilities
-  - openff-models
+  - rdkit !=2024.03.6,!=2024.09.1
```
Comment (mattwthompson) [inline devtools/conda-envs/test_env.yaml:15]: I was getting some local failures and copied the toolkit's pin, I can take this or leave it

File: smirnoff_plugins/collections/nonbonded.py [hunk 2]
```diff
@@ -532,7 +537,7 @@ def modify_openmm_forces(
                 force.addExclusion(i, j)
 
         elif len(existing_custom_nonbondeds) > 0:
-            nonbonded: openmm.CustomNonbondedForce = existing_custom_nonbondeds[0]
+            nonbonded: openmm.CustomNonbondedForce = existing_custom_nonbondeds[0]  # type: ignore[no-redef]
```
Comment (Yoshanuikabundi) [inline smirnoff_plugins/collections/nonbonded.py:540]: Removing the type hint here and on line 534 seems to allow you to remove the `type: ignore`:

```diff
          if len(existing_nonbondeds) > 0:
-             nonbonded: openmm.NonbondedForce = existing_nonbondeds[0]
+             nonbonded = existing_nonbondeds[0]
              for idx in range(nonbonded.getNumExceptions()):
                  i, j, _, _, _ = nonbonded.getExceptionParameters(idx)
                  force.addExclusion(i, j)
  
          elif len(existing_custom_nonbondeds) > 0:
-             nonbonded: openmm.CustomNonbondedForce = existing_custom_nonbondeds[0]  # type: ignore[no-redef]
+             nonbonded = existing_custom_nonbondeds[0]
```

**Review by Yoshanuikabundi — APPROVED** [review]
This looks good! I've squeezed the Buckingham water example into a notebook (see below) and it runs fine and the trajectory looks correct (or at least reasonable).

My notes are just using the actual release rather than the release candidate in the Conda environment and a few opportunities to remove `type: ignore` comments.

It's most likely out of scope for this PR, but I think the `smirnoff_plugins.utilities.openmm.simulate()` function could be simplified a lot with modern Interchange API ergonomics.

Notebook: [buckingham-water.tar.gz](https://github.com/user-attachments/files/17863690/buckingham-water.tar.gz)

File: smirnoff_plugins/collections/nonbonded.py (hunk 2 above)
Comment (mattwthompson) [reply smirnoff_plugins/collections/nonbonded.py:540]: Hmm yeah, I'm not sure why that was there in the first place

**Comment by mattwthompson** [comment]
Thanks @Yoshanuikabundi! Sorry for leaving some of the cruft in

**Pull request merged**